File: haskell/step8_macros.hs
import System.Environment (getArgs)
import Control.Monad (when, mapM)
import Control.Monad.Error (throwError)
import qualified Data.Map as Map
import qualified Data.Traversable as DT

import Readline (readline, load_history)
import Types
import Reader (read_str)
import Printer (_pr_str)
import Env (Env, env_new, env_bind, env_find, env_get, env_set)
import Core as Core

-- read
mal_read :: String -> IO MalVal
mal_read str = read_str str

-- eval
is_pair (MalList x:xs) = True
is_pair (MalVector x:xs) = True
is_pair _ = False

quasiquote :: MalVal -> MalVal
quasiquote ast =
    case ast of
         (MalList (MalSymbol "unquote" : a1 : [])) -> a1
         (MalList (MalList (MalSymbol "splice-unquote" : a01 : []) : rest)) ->
            MalList [(MalSymbol "concat"), a01, quasiquote (MalList rest)]
         (MalVector (MalList (MalSymbol "splice-unquote" : a01 : []) : rest)) ->
            MalList [(MalSymbol "concat"), a01, quasiquote (MalVector rest)]
         (MalList (a0 : rest)) -> MalList [(MalSymbol "cons"),
                                           quasiquote a0,
                                           quasiquote (MalList rest)]
         (MalVector (a0 : rest)) -> MalList [(MalSymbol "cons"),
                                             quasiquote a0,
                                             quasiquote (MalVector rest)]
         _ -> MalList [(MalSymbol "quote"), ast]

is_macro_call :: MalVal -> Env -> IO Bool
is_macro_call (MalList (a0@(MalSymbol _) : rest)) env = do
    e <- env_find env a0
    case e of
         Just e -> do
            f <- env_get e a0
            case f of
                 MalFunc {macro=True} -> return True
                 _                    -> return False
         Nothing -> return False
is_macro_call _ _ = return False

macroexpand :: MalVal -> Env -> IO MalVal
macroexpand ast@(MalList (a0 : args)) env = do
    mc <- is_macro_call ast env
    if mc then do
        mac <- env_get env a0
        case mac of 
             MalFunc {fn=(Fn f)} -> do
                new_ast <- f args
                macroexpand new_ast env
             _ ->
                return ast
    else
        return ast
macroexpand ast _ = return ast


eval_ast :: MalVal -> Env -> IO MalVal
eval_ast sym@(MalSymbol _) env = env_get env sym
eval_ast ast@(MalList lst) env = do
    new_lst <- mapM (\x -> (eval x env)) lst
    return $ MalList new_lst
eval_ast ast@(MalVector lst) env = do
    new_lst <- mapM (\x -> (eval x env)) lst
    return $ MalVector new_lst
eval_ast ast@(MalHashMap lst) env = do
    new_hm <- DT.mapM (\x -> (eval x env)) lst
    return $ MalHashMap new_hm
eval_ast ast env = return ast

let_bind :: Env -> [MalVal] -> IO Env
let_bind env [] = return env
let_bind env (b:e:xs) = do
    evaled <- eval e env
    x <- env_set env b evaled
    let_bind env xs

apply_ast :: MalVal -> Env -> IO MalVal
apply_ast ast@(MalList (MalSymbol "def!" : args)) env = do
    case args of
         (a1@(MalSymbol _): a2 : []) -> do
            evaled <- eval a2 env
            env_set env a1 evaled
         _ -> error $ "invalid def!"
apply_ast ast@(MalList (MalSymbol "let*" : args)) env = do
    case args of
         (MalList a1 : a2 : []) -> do
            let_env <- env_new $ Just env
            let_bind let_env a1
            eval a2 let_env
         (MalVector a1 : a2 : []) -> do
            let_env <- env_new $ Just env
            let_bind let_env a1
            eval a2 let_env
         _ -> error $ "invalid let*"
apply_ast ast@(MalList (MalSymbol "quote" : args)) env = do
    case args of
         a1 : [] -> return a1
         _ -> error $ "invalid quote"
apply_ast ast@(MalList (MalSymbol "quasiquote" : args)) env = do
    case args of
         a1 : [] -> eval (quasiquote a1) env
         _ -> error $ "invalid quasiquote"

apply_ast ast@(MalList (MalSymbol "defmacro!" : args)) env = do
    case args of
         (a1 : a2 : []) -> do
            func <- eval a2 env
            case func of
                MalFunc {fn=f, ast=a, env=e, params=p} -> do
                    let new_func = MalFunc {fn=f, ast=a, env=e,
                                            params=p, macro=True} in
                        env_set env a1 new_func
                _ -> error $ "defmacro! on non-function"
         _ -> error $ "invalid defmacro!" 
apply_ast ast@(MalList (MalSymbol "macroexpand" : args)) env = do
    case args of
         (a1 : []) -> macroexpand a1 env
         _ -> error $ "invalid macroexpand" 
apply_ast ast@(MalList (MalSymbol "do" : args)) env = do
    case args of
         ([]) -> return Nil
         _  -> do
            el <- eval_ast (MalList args) env
            case el of
                 (MalList el) -> return $ last el
            
apply_ast ast@(MalList (MalSymbol "if" : args)) env = do
    case args of
         (a1 : a2 : a3 : []) -> do
            cond <- eval a1 env
            if cond == MalFalse || cond == Nil
                then eval a3 env
                else eval a2 env
         (a1 : a2 : []) -> do
            cond <- eval a1 env
            if cond == MalFalse || cond == Nil
                then return Nil
                else eval a2 env
         _ -> error $ "invalid if"
apply_ast ast@(MalList (MalSymbol "fn*" : args)) env = do
    let params = case args of
                     ((MalList lst) : _)   -> lst
                     ((MalVector lst) : _) -> lst in
        case args of
             (a1 : a2 : []) -> do
                return $ (_malfunc a2 env a1 (\args -> do
                    fn_env1 <- env_new $ Just env
                    fn_env2 <- (env_bind fn_env1 params args)
                    eval a2 fn_env2))
             _ -> error $ "invalid fn*"
apply_ast ast@(MalList _) env = do
    mc <- is_macro_call ast env
    if mc then do
        new_ast <- macroexpand ast env
        eval new_ast env
    else
        case ast of
            MalList _ -> do
                el <- eval_ast ast env
                case el of
                    (MalList (Func (Fn f) : rest)) ->
                        f $ rest
                    (MalList (MalFunc {ast=ast,
                                    env=fn_env,
                                    params=(MalList params)} : rest)) -> do
                        fn_env1 <- env_new $ Just fn_env
                        fn_env2 <- (env_bind fn_env1 params rest)
                        eval ast fn_env2
                    el ->
                        error $ "invalid apply: " ++ (show el)
            _ -> return ast

eval :: MalVal -> Env -> IO MalVal
eval ast env = do
    case ast of
         (MalList lst) -> apply_ast ast env
         _             -> eval_ast ast env


-- print
mal_print :: MalVal -> String
mal_print exp = show exp

-- repl

rep :: Env -> String -> IO String
rep env line = do
    ast <- mal_read line
    exp <- eval ast env
    return $ mal_print exp

repl_loop :: Env -> IO ()
repl_loop env = do
    line <- readline "user> "
    case line of
        Nothing -> return ()
        Just "" -> repl_loop env
        Just str -> do
            out <- catchAny (rep env str) $ \e -> do
                return $ "Error: " ++ (show e)
            putStrLn out
            repl_loop env

main = do
    args <- getArgs
    load_history

    repl_env <- env_new Nothing

    -- core.hs: defined using Haskell
    (mapM (\(k,v) -> (env_set repl_env (MalSymbol k) v)) Core.ns)
    env_set repl_env (MalSymbol "eval") (_func (\[ast] -> eval ast repl_env))
    env_set repl_env (MalSymbol "*ARGV*") (MalList [])

    -- core.mal: defined using the language itself
    rep repl_env "(def! not (fn* (a) (if a false true)))"
    rep repl_env "(def! load-file (fn* (f) (eval (read-string (str \"(do \" (slurp f) \")\")))))"
    rep repl_env "(defmacro! cond (fn* (& xs) (if (> (count xs) 0) (list 'if (first xs) (if (> (count xs) 1) (nth xs 1) (throw \"odd number of forms to cond\")) (cons 'cond (rest (rest xs)))))))"
    rep repl_env "(defmacro! or (fn* (& xs) (if (empty? xs) nil (if (= 1 (count xs)) (first xs) `(let* (or_FIXME ~(first xs)) (if or_FIXME or_FIXME (or ~@(rest xs))))))))"

    if length args > 0 then do
        env_set repl_env (MalSymbol "*ARGV*") (MalList (map MalString (drop 1 args)))
        rep repl_env $ "(load-file \"" ++ (args !! 0) ++ "\")" 
        return ()
    else 
        repl_loop repl_env

File: dart/step8_macros.dart
#!/usr/bin/env dart

library mal.step8_macros;

import "dart:io";
import "types.dart";
import "reader.dart" as reader;
import "printer.dart" as printer;
import "env.dart";
import "core.dart" as core;

Eval eval = new Eval((arguments) => EVAL(arguments[0], repl_env));

isPair(MalType ast) => (ast is MalList || ast is MalVector) ? (ast as MalList).malTypes.isNotEmpty : false;

MalType quasiquote(MalType ast) {
  rest(MalList a) {
    if (a.malTypes.length > 0) {
      return new MalList.fromList(a.malTypes.sublist(1, a.malTypes.length));
    } else {
      return new MalList();
    }
  }

  if (!isPair(ast)) {
    // Not not of list type
    return new MalList.fromList([new MalSymbol("quote"), ast]);
  } else {

    MalType argument0 = (ast as MalList).malTypes[0];

    if ((argument0 is MalSymbol) && (argument0 as MalSymbol).symbol == "unquote") {
      return (ast as MalList).malTypes[1];
    } else if (isPair(argument0)) {
      MalType argument00 = (argument0 as MalList).malTypes[0];

      if ((argument00 is MalSymbol) && (argument00 as MalSymbol).symbol == "splice-unquote") {



        return new MalList.fromList([new MalSymbol("concat"), (argument0 as MalList).malTypes[1], quasiquote( rest(ast) )]);
      }
    }

    return new MalList.fromList([new MalSymbol("cons"), quasiquote(argument0), quasiquote( rest(ast) )]);
  }
}

MalType READ(String str) {
  return reader.read_str(str);
}

Env repl_env = new Env();

MalType eval_ast(MalType ast, Env env) {

  if (ast is MalSymbol) {
    return env.get(ast);
  } else if (ast is MalList) {

    MalList newMalList = ast is MalVector ? new MalVector() : new MalList();

    ast.malTypes.forEach((MalType malType) {
      MalType evaluatedMalType = EVAL(malType, env);
      newMalList.malTypes.add(evaluatedMalType);
    });

    return newMalList;
  } else if (ast is MalHashMap) {

    MalHashMap newMalHashMap = new MalHashMap();

    ast.malHashMap.forEach((key, value) {
      MalType evaluatedMalType = EVAL(value, env);
      newMalHashMap.malHashMap[key] = evaluatedMalType;
    });

    return newMalHashMap;
  } else {
    return ast;
  }
}

bool isMacroCall(MalType ast, Env env) {
  if (ast is MalList) {
    MalType argument0 = ast.nth(0);
    if (argument0 is MalSymbol && env.find(argument0) != null) {
      MalType macro = env.get(argument0);
      if (macro is MalFunction && macro.macro) {
        return true;
      }
    }

  }
  return false;
}

MalType macroExpand(MalType ast, Env env) {
  while(isMacroCall(ast, env)) {
    MalSymbol argument0 = ((ast as MalList).nth(0) as MalSymbol);
    MalFunction macro = (env.get(argument0) as MalFunction);
    ast = Function.apply(macro, (ast as MalList).rest().malTypes);
  }

  return ast;
}

MalType EVAL(MalType sourceAst, Env env) {
  var argument0, argument1, argument2, argument3;

  while(true) {
    if (!(sourceAst is MalList) || sourceAst is MalVector) {
      return eval_ast(sourceAst, env);
    }

    MalType expanded = macroExpand(sourceAst, env);
    if (!(expanded is MalList)) {
      return expanded;
    }

    sourceAst = expanded;

    if ((sourceAst as MalList).malTypes.length == 0) {
      return sourceAst;
    }

    var ast = sourceAst as MalList;
    argument0 = (ast.malTypes[0] is MalSymbol) ? (ast.malTypes[0] as MalSymbol).symbol : "__<*fn*>__";

    switch (argument0) {
      case 'def!':
        argument1 = ast.malTypes[1] as MalSymbol;
        argument2 = ast.malTypes[2];

        var result = EVAL(argument2, env);
        env.set(argument1, result);
        return result;
      case 'let*':
        argument1 = ast.malTypes[1] as MalList;
        argument2 = ast.malTypes[2];
        Env let_env = new Env(outer: env);

        for (int i = 0; i < argument1.malTypes.length; i += 2) {
          var key = argument1.malTypes[i] as MalSymbol;
          var value = argument1.malTypes[i + 1];
          let_env.set(key, EVAL(value, let_env));
        }

        sourceAst = argument2;
        env = let_env;
        break;
      case 'quote':
        return ast.malTypes[1];
      case 'quasiquote':
        sourceAst = quasiquote(ast.malTypes[1]);
        break;
      case 'defmacro!':
        argument1 = ast.nth(1);
        argument2 = ast.nth(2);
        var result = EVAL(argument2, env);
        (result as MalFunction).setMacro();
        env.set(argument1, result);
        return result;
      case 'macroexpand':
        argument1 = ast.nth(1);
        return macroExpand(argument1, env);
      case 'do':
      // 'do: return eval_ast(rest(ast), env)[-1]
      // * `do`: Evaluate the all the elements of the list using `eval_ast`
      //   and return the final evaluated element.
        rest(MalList a) {
          return new MalList.fromList(a.malTypes.getRange(1, a.malTypes.length - 1).toList());
        }

        eval_ast(rest(sourceAst), env) as MalList;
        sourceAst = ast.malTypes.last;
        break;
      case 'if':
      // 'if: return EVAL(EVAL(ast[1], env) ? ast[2] : ast[3], env)
      // * `if`: Evaluate the first parameter (second element). If the result
      //   (condition) is anything other than `nil` or `false`, then evaluate
      //   the second parammeter (third element of the list) and return the
      //   result.  Otherwise, evaluate the third parameter (fourth element)
      //   and return the result. If condition is false and there is no third
      //   parameter, then just return `nil`.
        argument1 = ast.malTypes[1];
        argument2 = ast.malTypes[2];

        var result = EVAL(argument1, env);

        if ((result is MalBoolean && (result as MalBoolean).value)) {

          sourceAst = argument2;
          break;
        }

        if (!(result is MalNil) && (result is MalType) && !(result is MalBoolean)) {
          sourceAst = argument2;
          break;
        }

        if (ast.malTypes.length < 4) {
          return MAL_NIL;
        }

        argument3 = ast.malTypes[3];
        sourceAst = argument3;
        break;
      case 'fn*':
      // 'fn*: return (...a) -> EVAL(ast[2], new Env(env, ast[1], a))
        argument1 = ast.malTypes[1] as MalList;
        argument2 = ast.malTypes[2];
        Env cur_env = env;
        return new MalFunction((List<MalType> arguments) =>
          EVAL(argument2, new Env(outer: cur_env, binds: argument1, exprs: new MalList.fromList(arguments))), ast: argument2, env: env, fParams: argument1);
      default:
        MalList astList = eval_ast(sourceAst, env);
        var function = astList.malTypes[0] as VarargsFunction;
        var args = astList.malTypes.getRange(1, astList.malTypes.length).toList();

        var functionAst = function.ast;
        if (functionAst != null) {
          sourceAst = functionAst;
          env = new Env(outer: function.env, binds: function.fParams, exprs: new MalList.fromList(args));
        } else {
          var result = Function.apply(function, args);
          return result;
        }
        break;
    }
  }
}

void PRINT(MalType exp) {
  stdout.writeln(printer.pr_str(exp));
}

void rep(String str) {
  PRINT(EVAL(READ(str), repl_env));
}

void init([List<String> args = const []]) {
  // Setup environment
  core.ns.forEach((key, value) => repl_env.set(new MalSymbol(key), value));
  repl_env.set(new MalSymbol("eval"), eval);
  repl_env.set(new MalSymbol("*ARGV*"), new MalList.fromList(args.map((e) => new MalString(e)).toList()));

  // Define global function in language
  rep("(def! not (fn* (a)(if a false true)))");
  rep('(def! load-file (fn* (f) (eval (read-string (str "(do" (slurp f) ")")))))');
  rep("(defmacro! cond (fn* (& xs) (if (> (count xs) 0) (list 'if (first xs) (if (> (count xs) 1) (nth xs 1) (throw \"odd number of forms to cond\")) (cons 'cond (rest (rest xs)))))))");
  rep("(defmacro! or (fn* (& xs) (if (empty? xs) nil (if (= 1 (count xs)) (first xs) `(let* (or_FIXME ~(first xs)) (if or_FIXME or_FIXME (or ~@(rest xs))))))))");
}

void main(List<String> args) {
  String line;

  init(args);

  if (args.length > 0) {
    rep('(load-file "${args[0]}")');
    exit(0);
  }

  while (true) {
    stdout.write("user> ");
    try {
      line = stdin.readLineSync();
      if (line == null) {
        // Control signal or EOF then break from REPL.
        break;
      }

      rep(line);
    } on Exception catch (ex) {
      stdout.writeln("Error: ${ex.toString()}");
    } on StateError catch (ex) {
      stdout.writeln("Error: ${ex.message}");
    } on StackOverflowError catch (ex) {
      stdout.writeln("Error: ${ex.toString()}");
      PRINT(MAL_NIL);
    }
  }
}
